### ISSUE TITLE
Type Union with NoReturn

### ISSUE THREAD

**Issue by jfmend** [microsoft/pylance-release#2008]
<h3> Environment data </h3>

-   Language Server version: 2021.10.3
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: basic

<h3> Expected behaviour </h3>

No type errors reported

<h3> Actual behaviour </h3>

Error resulting from `Union[A, B, ..., NoReturn]` not being considered equivalent to `Union[A, B, ...]` which I believe it should.
To be clear, a [variable, function_arg, return_type] of type `A` **should not accept** an expression of type `NoReturn` (since this is either useless or a bug) in which case a type error is correctly reported. However, a scenario where a type `A` is expected **should accept** an expression of type `Union[A, NoReturn]`.

Notice that, in Python, the evaluation of any expression may result in an `Exception` being raised (ex: `MemoryError`). This means that an expression of some type `A` is always implicitly of type `Union[A, NoReturn]`.

<h3> Code Snippet / Additional information </h3>

```python
from typing import NoReturn


def raises(exception: Exception) -> NoReturn:
    '''Helper function to allow raising an exception in an expression context'''
    raise exception


# Example 1 

n = 1
n_str: str
# incorrect type ERROR (assigning str | NoReturn to var of type str):
n_str = str(n) if n > 0 else raises(
        Exception(f'Must be positive: N = {n}'))
print(n_str)


# Example 2

import os
from pathlib import Path


def find_productFile(product_dir: Path) -> Path:
    dir_files = os.listdir(product_dir)
    # incorrect type ERROR (returning Path | NoReturn where Path is expected):
    return ( product_dir/'product.xml'   if 'product.xml' in dir_files
        else product_dir/'manifest.safe' if 'manifest.safe' in dir_files
        else raises(Exception(
                'Product not identified as one of [Sentinel-1, RadarSat2]')))
```

**Comment by jfmend** [comment]
Thank you for the quick reply.
I am not sure if this should be another issue or not, but type check errors with `Union`s containing `NoReturn` also occurs in `or` expressions:
```python
from typing import NoReturn


def raises(exception: Exception) -> NoReturn:
    '''Helper function to allow raising an exception in an expression context'''
    raise exception


n = 1
n_str: str
# incorrect type ERROR (assigning str | NoReturn to var of type str):
n_str = str(n) or raises(RuntimeError(f'Must be positive: N = {n}'))
print(n_str)
```

**Comment by erictraut** [comment]
I've added support for `or` and `and` operators as well.

**Comment by heejaechang** [comment]
fixed in 2021.11.0
